### PR TITLE
Make dataclass compatible with schema

### LINKED ISSUE #792: AttributeError when using dataclass with schema function
# Bug

Please complete:
* OS: macOS
* Python version `import sys; print(sys.version)`: 3.7.4
* Pydantic version `import pydantic; print(pydantic.VERSION)`: 0.32.2, as well as master as of 7901711

Where possible please include a self contained code snippet describing your bug:

```py
from pydantic.dataclasses import dataclass
from pydantic.schema import schema

@dataclass
class Model:
    name: str

schema([Model])
```

This produces an error message like:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jamie/.virtualenvs/tempenv-74cd74727d61/lib/python3.7/site-packages/pydantic/schema.py", line 185, in schema
    flat_models = get_flat_models_from_models(models)
  File "/Users/jamie/.virtualenvs/tempenv-74cd74727d61/lib/python3.7/site-packages/pydantic/schema.py", line 437, in get_flat_models_from_models
    flat_models |= get_flat_models_from_model(model)
  File "/Users/jamie/.virtualenvs/tempenv-74cd74727d61/lib/python3.7/site-packages/pydantic/schema.py", line 378, in get_flat_models_from_model
    fields = cast(Sequence[Field], model.__fields__.values())
AttributeError: type object 'Model' has no attribute '__fields__'
```

The same code using `BaseModel` instead of the `@dataclass` works fine.

With the `dataclass`, `Model.__pydantic_model__.schema()` also works.

### PR DESCRIPTION
**Pull request by aviramha**
## Change Summary

Add compatibility for schema and schema_model so it can be used easily on a dataclass.
Also add a proxy method for dataclass of schema.

## Related issue number

fix #792 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

It's a pretty hacky way, but it's also simple and easy.
I know I'm failing on mypy run but am not sure how to handle this (not so familiar with mypy). Advice is welcome.

### CODE REVIEW

File: pydantic/dataclasses.py [hunk 1]
```diff
@@ -115,6 +115,7 @@ def _pydantic_post_init(self: 'DataclassType', *initvars: Any) -> None:
     if cls.__pydantic_model__.__config__.validate_assignment and not frozen:
         cls.__setattr__ = setattr_validate_assignment
 
+    cls.schema = cls.__pydantic_model__.schema
```
Comment (samuelcolvin) [inline pydantic/dataclasses.py]: please remove this.

File: changes/792-aviramha.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+* Change schema and schema_model to handle dataclasses by using their __pydantic_model__ feature.
```
Comment (samuelcolvin) [inline changes/792-aviramha.md]:
```suggestion
* Change schema and schema_model to handle dataclasses by using their `__pydantic_model__` feature.
```
Also, this should be one line and not include stars at the beginning.

File: pydantic/dataclasses.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/dataclasses.py]: see #481 for how we should implement methods like this for dataclasses.

File: pydantic/utils.py [hunk 3]
```diff
@@ -114,6 +114,16 @@ def almost_equal_floats(value_1: float, value_2: float, *, delta: float = 1e-8)
     return abs(value_1 - value_2) <= delta
 
 
+def is_basemodel_subclass(cls: AnyType) -> bool:
```
Comment (samuelcolvin) [inline pydantic/utils.py]: I think better to implement

```py
def get_model(obj: Union[Type['BaseModel'], Type['DataclassType']]) -> Type['BaseModel']:
    from .main import BaseModel
    try:
        model_cls = obj.__pydantic_model__
    except AttributeError:
        model_cls = obj
    
    if not issubclass(model_cls, BaseModel):
        raise TypeError('...')
    return model_cls
```

**Comment by samuelcolvin** [comment]
tests are failing. Best if you can run tests and linting with just `make` (it's very quick) before pushing.

File: changes/792-aviramha.md (hunk 2 above)
Comment (samuelcolvin) [reply changes/792-aviramha.md]: You've still missed my suggested change here.

**Comment by aviramha** [comment]
> tests are failing. Best if you can run tests and linting with just `make` (it's very quick) before pushing.

I'm aware of that and wrote in the original post. I'm not quite sure how fix it as it's myoy failure.

**Comment by samuelcolvin** [comment]
You need to import `DataclassType` (perhaps within `if TYPE_CHECKING:`)

**Comment by samuelcolvin** [comment]
`pydantic/utils.py:121` you can use `# type: ignore`

File: changes/792-aviramha.md (hunk 2 above)
Comment (aviramha) [reply changes/792-aviramha.md]: Ah, sorry. Fixed it now :)

**Comment by aviramha** [comment]
Thanks for the review. I hope the current changes matches your expectations :)

**Comment by samuelcolvin** [comment]
thanks so much.

**Comment by aviramha** [comment]
> thanks so much.

Thanks for the awesome library!!

**Pull request merged**